src/docs/autoscaling.md

```diff
@@ -0,0 +1,115 @@
+
+For example, 
+```notify-delivery-worker-sender``` is an app for delivering the notifications to the service providers, i.e. AWS SES and text messaging providers. The celery worker consumes job from two queues: ```send-sms-tasks``` and ```send-sms-tasks``` are being monitored. The number of instances is then calculated as:-
+
+```
+desired_instances_count = total_message_count / request_per_instance
+```
+where `total_message_count` is the sum of all the queues of this app and `request_per_instance` is a predefined estimated number of jobs an app instance can handle at one time.
+
+### Request handlers ##
+
+The inbound apps are scaled based on last 5 `request_counts`, where `request_counts` is the total number of incoming requests in one minute.  
+
+```
+desired_instance_count = max(request_counts)/request_per_instance
+```
+
+`request_per_instance` is a pre-defined scaler that indicates the number of requests that can be handled by one instance of app. 
+
+
+### Scheduled jobs ##
+Notify uses a proactive schedule for scheduled tasks. Scheduled tasks are CSV files the users uploaded to be run at a given time. We already have the information of job size (number of notificaitons) and the scheduled time. The proactive scheduler scales up the number of instances before the jobs start. This scaling applies to the same task as the Delivery workers above. 
+
+The desired number of instances for the particular app is calculated as follow:
+
+```
+desired_instance_count=num_of_scheduled_jobs/request_per_instance/scheduled_job_scaling_factor
+and
+min_instance_count<=desired_instances_count<=max_instance_count
+```
+
+The `num_of_scheduled_jobs` is the number of notifications to be sent in that task. `Request_per_instance` gives a first indication on the number of jobs that can be handled by one app instance. We can use `scheduled_job_scaling_factor ` to control how aggressively we want the scaling to be. Currently, this factor is set to 2 in Notify. Moreover, as the jobs are first delivered to the queues by delivery worker before these queues are consumed by senders and db workers. Hence, they are scaled up less agressively. 
+
+## The Scaling function ##
+The function `scale_paas_apps` uses the `desired_instance_count` factor calculated to scale up or down the number of instances for the app. Below is the Psuedocode for the autoscaler. 
+
+```
+if desired_instance_count > current_instance_count
+	scale up to desired_instance_count
+else if time since last scale up > 5 mins
+	current_instance_count = current_instance_count - 1
+```
+
+The apps are instantly scaled up to the `desired_instance_count` if the current number of running instances is lower. However, on scaling down, No apps are scaled down if there has been a scaling up event in any app in the last 5 mins.  
```

Comment:
However, a scale down is only possible if more than 5 minutes have passed since the last scale up event for each particular app.